[JSC] Improve `Math.hypot` Performance with 4+ Arguments

Gumichocopengin8 opened the pull request:
#### 60dfa5abdbb3ef5b9cf73040da71a5b3378ab876
<pre>
[JSC] Improve `Math.hypot` Performance with 4+ Arguments
<a href="https://bugs.webkit.org/show_bug.cgi?id=301480">https://bugs.webkit.org/show_bug.cgi?id=301480</a>

Reviewed by Keith Miller.

This patch improves `Math.hypot` performance with 4+ arguments.

                                TipOfTree                   Patched                              Ratio
math-hypot-common          0.7415+-0.2021     ?      0.8036+-0.0676        ? might be 1.0838x slower
math-hypot-uncommon        1.4897+-0.0405     ^      0.9222+-0.1222        ^ definitely 1.6154x faster
math-hypot-rare            1.9600+-0.0540     ^      1.0715+-0.1103        ^ definitely 1.8293x faster
math-hypot-epic            2.8179+-0.1830     ^      1.1715+-0.1121        ^ definitely 2.4054x faster

Tests: JSTests/microbenchmarks/math-hypot-common.js
       JSTests/microbenchmarks/math-hypot-epic.js
       JSTests/microbenchmarks/math-hypot-rare.js
       JSTests/microbenchmarks/math-hypot-uncommon.js
       JSTests/stress/math-hypot.js

* JSTests/microbenchmarks/math-hypot-common.js: Added.
(hypot2):
(hypot3):
* JSTests/microbenchmarks/math-hypot-epic.js: Added.
(hypot8):
(hypot9):
* JSTests/microbenchmarks/math-hypot-rare.js: Added.
(hypot6):
(hypot7):
* JSTests/microbenchmarks/math-hypot-uncommon.js: Added.
(hypot4):
(hypot5):
* JSTests/microbenchmarks/math-hypot.js: Removed.
* JSTests/stress/math-hypot.js: Added.
(shouldBe):
(i.generateObj):
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/302558@main">https://commits.webkit.org/302558@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ef25f2c5884b3523b7e5d1cde7c31892a165582f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129406 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136783 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80817 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1594 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1539 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98556 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66448 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1245 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115906 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79207 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34037 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80060 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121398 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109622 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34537 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139257 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127858 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1400 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107082 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112250 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106926 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1184 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30770 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/54119 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20207 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1524 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64887 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160872 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1341 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40126 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1378 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1446 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Gumichocopengin8 commented on JSTests/stress/math-hypot.js:
```diff
@@ -0,0 +1,186 @@
+        1.5819637896591838e308
+    );
+    shouldBe(Math.hypot(-2.534858246857893e115, 8.988465674311579e307, 8.98846567431158e307), 1.2711610061536462e308);
+    // FIXME: This test result differs from Chrome and Firefox,
+    // although it matches the behavior of Swift, C++, Rust, and Python. Commenting out this test case for now.
+    // shouldBe(Math.hypot(1.3588124894186193e308, 1.4803986201152006e223, 6.741349255733684e307), 1.5168484694516577e308);
```
> [!NOTE]
> This patch does not change the outcome of this test.

kmiller68 commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -246,27 +246,35 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncHypot, (JSGlobalObject* globalObject, Call
+    bool hasNaN = false;
+    double maxAbs = 0;
+    double sum = 0;
     for (double argument : args) {
```
If we wanted to make this faster, I would expect it's faster to do the `toNumber()` inside this loop. Nothing in this loop has observable effects and throwing is likely rare. That avoids the vector entirely.

Gumichocopengin8 replied:
Fixed it. That improved the performance even more!

Gumichocopengin8 replied:
Actually `toNumber()` has observable effects, so used a flag instead of the early return with `Infinity` case.
Also, updated the test cases to cover this case.

kmiller68 commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -246,27 +246,35 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncHypot, (JSGlobalObject* globalObject, Call
+    if (!maxAbs) [[unlikely]]
+        return JSValue::encode(jsDoubleNumber(0.0));
+    if (hasNaN) [[unlikely]]
+        return JSValue::encode(jsNaN());
```
These seem backwards and seem like a behavioral change (e.g. today `Math.hypot(0.0, NaN) === NaN`). Can you add a test?

Gumichocopengin8 replied:
Ah the order was wrong. Fixed it.

kmiller68 commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -246,27 +246,35 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncHypot, (JSGlobalObject* globalObject, Call
+            double scaledArgument = maxAbs / absArgument;
+            sum = std::fma(sum, scaledArgument * scaledArgument, 1);
+            maxAbs = absArgument;
+        } else if (0 < maxAbs) {
```
When can this case happen?

Gumichocopengin8 replied:
This happens with cases like `Math.hypot(4, 3, 2, 1)` where the current value is smaller than `maxAbs`.

kmiller68 replied:
Sorry, I read it as `maxAbs < 0` which seemed like an invariant. Wouldn't this be clearer as `if (maxAbs)` though?

Gumichocopengin8 replied:
Fixed!

kmiller68 commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -246,27 +246,35 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncHypot, (JSGlobalObject* globalObject, Call
+        if (std::isinf(argument)) [[unlikely]]
             return JSValue::encode(jsDoubleNumber(+std::numeric_limits<double>::infinity()));
-        max = std::max(std::abs(argument), max);
+        if (std::isnan(argument)) [[unlikely]] {
+            hasNaN = true;
+            continue;
```
Why not just early return here? If you make the above change and get rid of the Vector then you'll need to run through the rest of the loop though.

Gumichocopengin8 replied:
Because `Infinity` has higher priority than `NaN` to return. `Math.hypot(NaN, NaN, NaN, Infinity)` should return `Infinity`.

kmiller68 reviewed (approved):
r=me

Gumichocopengin8 committed:
[JSC] Improve `Math.hypot` Performance with 4+ Arguments
https://bugs.webkit.org/show_bug.cgi?id=301480

Reviewed by Keith Miller.

This patch improves `Math.hypot` performance with 4+ arguments.

                                TipOfTree                   Patched                              Ratio
math-hypot-common          0.7415+-0.2021     ?      0.8036+-0.0676        ? might be 1.0838x slower
math-hypot-uncommon        1.4897+-0.0405     ^      0.9222+-0.1222        ^ definitely 1.6154x faster
math-hypot-rare            1.9600+-0.0540     ^      1.0715+-0.1103        ^ definitely 1.8293x faster
math-hypot-epic            2.8179+-0.1830     ^      1.1715+-0.1121        ^ definitely 2.4054x faster

Tests: JSTests/microbenchmarks/math-hypot-common.js
       JSTests/microbenchmarks/math-hypot-epic.js
       JSTests/microbenchmarks/math-hypot-rare.js
       JSTests/microbenchmarks/math-hypot-uncommon.js
       JSTests/stress/math-hypot.js

* JSTests/microbenchmarks/math-hypot-common.js: Added.
(hypot2):
(hypot3):
* JSTests/microbenchmarks/math-hypot-epic.js: Added.
(hypot8):
(hypot9):
* JSTests/microbenchmarks/math-hypot-rare.js: Added.
(hypot6):
(hypot7):
* JSTests/microbenchmarks/math-hypot-uncommon.js: Added.
(hypot4):
(hypot5):
* JSTests/microbenchmarks/math-hypot.js: Removed.
* JSTests/stress/math-hypot.js: Added.
(shouldBe):
(i.generateObj):
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/302558@main
